docs/proposals/horizontal-pod-autoscaler.md

```diff
@@ -0,0 +1,272 @@
+	// The number may be different because pod downscaling is someteimes delayed to keep the number
+	// of pods stable.
+	DesiredReplicas int
+
+	// CurrentConsumption is the current average consumption of the given resource that the autoscaler will
+	// try to maintain by adjusting the desired number of pods.
+	// Two types of resources are supported: "cpu" and "memory".
+	CurrentConsumption ResourceConsumption
+
+	// LastScaleTimestamp is the last time the HorizontalPodAutoscaler scaled the number of pods.
+	// This is used by the autoscaler to controll how often the number of pods is changed.
+	LastScaleTimestamp *util.Time
+}
+
+// ResourceConsumption is an object for specifying average resource consumption of a particular resource.
+type ResourceConsumption struct {
+	Resource api.ResourceName
+	Quantity resource.Quantity
+}
+```
+
+```Scale``` will be a reference to the Scale subresource.
+```MinCount```, ```MaxCount``` and ```Target``` will define autoscaler configuration.
+We will also introduce HorizontalPodAutoscalerList object to enable listing all autoscalers in the cluster:
+
+```go
+// HorizontalPodAutoscaler is a collection of pod autoscalers.
+type HorizontalPodAutoscalerList struct {
+	api.TypeMeta
+	api.ListMeta
+
+	Items []HorizontalPodAutoscaler
+}
+```
+
+## Autoscaling Algorithm
+
+The autoscaler will be implemented as a control loop.
+It will periodically (e.g.: every 1 minute) query pods described by ```Status.PodSelector``` of Scale subresource,
+and check their average CPU or memory usage from the last 1 minute
+(there will be API on master for this purpose, see
+[#11951](https://github.com/GoogleCloudPlatform/kubernetes/issues/11951).
+Then, it will compare the current CPU or memory consumption with the Target,
+and adjust the count of the Scale if needed to match the target
+(preserving condition: MinCount <= Count <= MaxCount).
+
+The target number of pods will be calculated from the following formula:
+
+```
+TargetNumOfPods = sum(CurrentPodsConsumption) / Target
+```
+
+To make scaling more stable, scale-up will happen only when the floor of ```TargetNumOfPods``` is higher than
+the current number, while scale-down will happen only when the ceiling of ```TargetNumOfPods``` is lower than
+the current number.
+
+The decision to scale-up will be executed instantly.
+However, we will execute scale-down only if the sufficient time has past from the last scale-up (e.g.: 10 minutes).
+Such approach has two benefits:
+
+* Autoscaler works in a conservative way.
+  If new user load appears, it is important for us to rapidly increase the number of pods,
+  so that user requests will not be rejected.
+  Lowering the number of pods is not that urgent.
+
+* Autoscaler avoids thrashing, i.e.: prevents rapid execution of conflicting decision if the load is not stable.
+
+
+As the CPU consumption of a pod immediately after start may be highly variable due to initialization/startup,
```

Comment:
In practice we've had two concrete memory consumption based scalers -
sharded redis clusters, and sharded frontends that are keeping in memory
stores.  However, typically it requires them have some way to allocate
keys, so it's practically more useful for cassandra or redis than for web
frontends.

On Wed, Aug 19, 2015 at 1:19 AM, David Oppenheimer <notifications@github.com
> wrote:

> In docs/proposals/horizontal-pod-autoscaler.md
> <https://github.com/kubernetes/kubernetes/pull/12859#discussion_r37380628>
> :
>
> > +the current number, while scale-down will happen only when the ceiling of ```TargetNumOfPods``` is lower than
> > +the current number.
> > +
> > +The decision to scale-up will be executed instantly.
> > +However, we will execute scale-down only if the sufficient time has past from the last scale-up (e.g.: 10 minutes).
> > +Such approach has two benefits:
> > +
> > +* Autoscaler works in a conservative way.
> > +  If new user load appears, it is important for us to rapidly increase the number of pods,
> > +  so that user requests will not be rejected.
> > +  Lowering the number of pods is not that urgent.
> > +
> > +* Autoscaler avoids thrashing, i.e.: prevents rapid execution of conflicting decision if the load is not stable.
> > +
> > +
> > +As the CPU consumption of a pod immediately after start may be highly variable due to initialization/startup,
>
> Related to my previous comment... Some apps will grow in memory
> consumption over time due to memory leaks. They may never stabilize. And in
> fact the increase in memory consumption may be solely due to memory leaks,
> and unrelated to load. I guess we would just tell people running such apps
> not to use memory-based autoscaling...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12859/files#r37380628>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift